### ISSUE TITLE
Unable to implement VBA.Collection

### ISSUE THREAD

**Issue by bclothier** [WaynePhillipsEA/twinbasic#164]
**Describe the bug**
If we try to implement an interface that contains members with unknown types, the implementation will remain incomplete. In this example, we cannot implement:
```
Public Function [_NewEnum]() As Unknown Implements VBA.Collection.[_NewEnum]
End Function
```

Because `Unknown` is not a recognized type. In this case, it should have worked:
```
Public Function [_NewEnum]() As IUnknown Implements VBA.Collection.[_NewEnum]
End Function
```

Which is a valid data type but fails to match the definition of the `VBA.Collection.[_NewEnum]`.

It should be noted that even in the case of an implementation requiring an object type that is not recognized, `IUnknown` should be always be compatible even if the original type library calls for something different.

**To Reproduce**
1. Create a class and try to implement the `VBA.Collection.
2. Insert the code:
```
Public Class Test
	Implements VBA.Collection
	
	Public Function Item(ByRef Index As Variant) As Variant Implements VBA.Collection.Item
	End Function
	
	Public Sub Add(ByRef Item As Variant, Optional ByRef Key As Variant, Optional ByRef Before As Variant, Optional ByRef After As Variant) Implements As VBA.Collection.Add
	End Sub
	
	Public Property Get Count() As Long Implements VBA.Collection.Count
	End Property
	
	Public Sub Remove(ByRef Index As Variant) Implements VBA.Collection.Remove
	End Sub
	
	Public Function [_NewEnum]() As IUnknown Implements VBA.Collection.[_NewEnum]
	End Function
End Class
```
3. Observe problems being reported on the implementation of `_NewEnum` as depicted in the screenshot:
![image](https://user-images.githubusercontent.com/2367644/119915939-29283f80-bf29-11eb-81d7-0e07d6c0e7c4.png)

4. Try to change the `IUnknown` to `Unknown` to match the description. This causes a different error as this is not a recognized data type.

**Expected behavior**
`IUnknown` should be always a valid option for unknown object data types and even if it is known, it still should be possible to substitute the explicit interface with `IUnknown` interface. 

**Desktop (please complete the following information):**
 - OS: Windows 7
 - twinBASIC compiler version 0.10.313

**Comment by cristianbuse** [comment]
Apologies, this is not related to be issue raised by @bclothier but it became obvious only once I saw his code. The following code is valid in tB:  
```VBA
Module Test
    Public Sub Main()
        Msgbox _Name
    End Sub
	
	Public Function _Name() As String
		Return "test"
	End Function
End Module
```
but method names are not allowed to start with underscore in VBA (at least).
Is it worth opening a new issue for this?

**Comment by WaynePhillipsEA** [comment]
@cristianbuse I'm inclined to allow that since it wont break backwards compatibility.  Though in 'legacy mode' where we want to strive for full compliance with VB6/VBA, we should actively disallow it.

**Comment by WaynePhillipsEA** [comment]
> IUnknown should be always a valid option for unknown object data types and even if it is known, it still should be possible to substitute the explicit interface with IUnknown interface.

No, this is not correct.  Let's take a return type as an example, if the definition gives say a return type of Access.Application, and we allow you to define it as IUnknown in your class, you could write an implementation that returns an Excel.Application object instead.   The caller **_is not_** required to do QueryInterface() on the returned IUnknown, and is allowed to assume that the caller returned an object that is of the defined type.  In doing so, the caller can now call directly into the vtable given by the returned object, but of course the vtable members are not correct because we got an Excel.Application vtable instead of the expected Access.Application vtable.   If you're lucky, you'll get an error.  If you're not so lucky, you'll get a hard crash, or you'll corrupt the stack and later get a hard crash.    The exact same problem goes for ByRef parameters.

In theory, you could allow IUnknown to be a substitute for ByVal object parameters, but since it's always better to have the proper defined type, I see no good reason to change this.

In this particular case, it is the slightly quirky internal representation of the VBA._Collection interface causing the 'unable to match' error.  I've got some work to do on VBA.Collection in #153, so I'll fix this soon.

**Comment by bclothier** [comment]
Please be aware this report is not limited to the `VBA.Collection`. I first observed it on a enumerable DAO type and used `VBA.Collection` as example in my bug report.

In both types’ cases, the IDL reports that they return an `IUnknown`, not the expected IEnumVARIANT` that would be used for enumerating. I don’t think I have seen any definitions in common type libraries actually return an `IEnumVARIANT`.

It is my understanding that iIn VB*, a `IUnknown` variable cannot be invoked upon, only passed around. Late bound calls on such variables aren’t allowed; one would have to cast to `Object` (aka `IDispatch`) to do late bound calls. Calling methods on the vtable requires more work on the programmer’s part and thus would be a very deliberate choice. If tB restricts late bound call on an `IUnknown` variable, then there should be no problem, right?

**Comment by WaynePhillipsEA** [comment]
> In both types’ cases, the IDL reports that they return an IUnknown, not the expected IEnumVARIANT that would be used for enumerating. I don’t think I have seen any definitions in common type libraries actually return an IEnumVARIANT.

The reason you can do that is because the NewEnum member is actually invoked through IDispatch, using the DISPID_NEWNUM (-4) member ID, with the side effect that all parameters and return value then pass through IDispatch::Invoke as Variants.  When passing objects through Variants, you 'lose' the exact type information, and the interfaces are then treated as either IUnknown or IDispatch, so the caller to the NewEnum member (through IDispatch::Invoke), HAS to call QueryInterface in order to obtain the actual IEnumVARIANT interface.

The problem is that in COM we're rarely implementing a pure dispinterface.  More usually we're implementing a dual interface, and so we HAVE to use the exact types in order to be COM compliant, for the reasons already mentioned in my previous post.  It's not a problem of tB restricting access to the IUnknown object... it's a problem of the caller, which could be C++, expecting you to return the exact correct type, with the correct vtable, since it can make early-bound / IUnknown calls on the returned object if it wishes.

Going back to the problem at hand, I expect the internal IUnknown used by tB is not being matched up to the stdole.IUnkown for the validation check being done by Implements.  It should be an easy fix when I get to it.

**Comment by WaynePhillipsEA** [comment]
Fixed in v0.10.833, thanks for reporting.

**Comment by bclothier** [comment]
This is fixed but there is one small UX inconsistency. The intellisense/problem console still shows `Unknown` when it probably should indicate `IUnknown`. 
![image](https://user-images.githubusercontent.com/2367644/120679013-4fd10380-c45e-11eb-8ecd-02464ddb266a.png)
